### PR TITLE
LiveSync - link to categories and sessions

### PR DESCRIPTION
**Pull request by pferreir**
Simplified the way records are processed.

See also https://github.com/indico/indico/pull/2545.

Missing:
* [x] Unit tests (need some dummy DB objects)
* [x] Checking that Invenio/CERN plugins work fine
* [x] Some extra smoke testing (handling deletes, etc...)

### CODE REVIEW

File: livesync/indico_livesync/handler.py [hunk 1]
```diff
@@ -88,12 +79,17 @@ def connect_signals(plugin):
     plugin.connect(signals.attachments.attachment_updated, _attachment_changed)
 
 
-def _moved(obj, old_parent, new_parent, **kwargs):
+def _moved_event(obj, old_category, **kwargs):
```
Comment (OmeGak) [inline livesync/indico_livesync/handler.py]: It's not clear what this function does unless you go into `_moved`. What about a more descriptive name?

Comment (ThiefMaster) [inline livesync/indico_livesync/handler.py]: Does it really matter though? It's a signal handler, not some API.
Anyway, It looks like the signal could be connected directly to `_moved`

File: livesync/indico_livesync/simplify.py [hunk 2]
```diff
@@ -16,10 +16,17 @@
 
 from __future__ import unicode_literals
 
+import itertools
 from collections import defaultdict
+from sqlalchemy.orm import joinedload
```
Comment (ThiefMaster) [inline livesync/indico_livesync/simplify.py]: that's not stdlib

File: livesync/indico_livesync/models/queue.py [hunk 3]
```diff
@@ -176,62 +209,72 @@ class LiveSyncQueueEntry(db.Model):
...
-        for agent in LiveSyncAgent.find():
-            for change in changes:
+        for change in changes:
+            obj = obj_deref(ref)
+            if isinstance(obj, Category):
+                if excluded_categories & {c.id for c in obj.chain_query}:
+                    continue
+            else:
+                event = obj if isinstance(obj, Event) else obj.event_new
+                if Event.query.filter(
```
Comment (OmeGak) [inline livesync/indico_livesync/models/queue.py]: Mixed multi-line style

File: livesync/indico_livesync/handler.py (hunk 1 above)
Comment (OmeGak) [inline livesync/indico_livesync/handler.py]: I know it's a handler, but I think it's generally better to name the functions after what they do and not after what calls them.

Comment (ThiefMaster) [inline livesync/indico_livesync/handler.py]: That does not make much sense in livesync - all of them eventually create livesync entries for an action

Comment (OmeGak) [inline livesync/indico_livesync/handler.py]: `_register_event_move` and nobody has to wonder.

Comment (pferreir) [inline livesync/indico_livesync/handler.py]: I'd say that since the convention with event handlers right now is to name the function after the event, we should keep this one consistent.
In this particular case, it would have to be `_register_move`, since this handler is also used by categories. Anyway, prefixing everything with `_register` seems like a bit of an overkill to me...

File: livesync/indico_livesync/simplify.py [hunk 4]
```diff
@@ -44,23 +53,80 @@ def process_records(records):
...
+    category_prot_records -= category_move_records  # A move already implies sending the whole record
+
+    # Protection changes are handled differently, as there may not be the need to re-generate the record
+    if category_prot_records:
+        for categ in Category.find(Category.id.in_(category_prot_records)):
+            cte = categ.get_protection_parent_cte()
+            # Update only children that inherit
+            inheriting_categ_children = (Event.query
+                                         .join(Category)
+                                         .join(cte, (Category.id == cte.c.id) & (cte.c.protection_parent == categ.id)))
```
Comment (ThiefMaster) [inline livesync/indico_livesync/simplify.py]: Do you actually need to `.join(Category)` here? Why not use `Event.category_id` in this filter instead

File: livesync/indico_livesync/util.py [hunk 5]
```diff
@@ -46,36 +47,24 @@ def obj_ref(obj):
...
+@memoize_request
 def obj_deref(ref):
     """Returns the object identified by `ref`"""
     from indico_livesync.models.queue import EntryType
     if ref['type'] == EntryType.category:
-        return CategoryManager().getById(ref['category_id'], True)
+        return Category.get(ref['category_id'])
     elif ref['type'] == EntryType.event:
-        return Event.get(ref['event_id'])
+        return Event.find(id=ref['event_id']).one()
```
Comment (ThiefMaster) [inline livesync/indico_livesync/util.py]: `Event.get_one()`?

File: livesync/indico_livesync/util.py [hunk 6]
```diff
@@ -46,36 +47,24 @@ def obj_ref(obj):
     return ImmutableDict(ref)
 
 
+@memoize_request
 def obj_deref(ref):
     """Returns the object identified by `ref`"""
     from indico_livesync.models.queue import EntryType
     if ref['type'] == EntryType.category:
-        return CategoryManager().getById(ref['category_id'], True)
+        return Category.get(ref['category_id'])
```
Comment (ThiefMaster) [inline livesync/indico_livesync/util.py]: `get_one` for consistency

File: livesync/indico_livesync/util.py [hunk 7]
```diff
@@ -46,36 +47,24 @@ def obj_ref(obj):
...
     elif ref['type'] == EntryType.event:
-        return Event.get(ref['event_id'])
+        return Event.find(id=ref['event_id']).one()
+    elif ref['type'] == EntryType.session:
+        return Session.find(id=ref['session_id']).one()
     elif ref['type'] == EntryType.contribution:
-        return Contribution.get(ref['contrib_id'])
+        return Contribution.find(id=ref['contrib_id']).one()
```
Comment (ThiefMaster) [inline livesync/indico_livesync/util.py]: `get_one()`

File: livesync/indico_livesync/util.py [hunk 8]
```diff
@@ -91,25 +80,18 @@ def clean_old_entries():
...
-    from indico_livesync.models.queue import EntryType
-    if ref['type'] == EntryType.category:
-        return ref['category_id'] in get_excluded_categories()
-    else:
-        obj = obj_deref(ref)
-        return unicode(obj.event_new.category_id) in {unicode(x) for x in get_excluded_categories()}
+def compound_id(obj):
+    """Generate a hierarchical compound ID, separated by dots."""
+    if isinstance(obj, (Category, Session)):
+        raise ValueError('Compound IDs are not supported for this entry type')
```
Comment (ThiefMaster) [inline livesync/indico_livesync/util.py]: `TypeError`?

File: livesync/tests/queue_test.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+
+def test_excluded_categories(mocker, monkeypatch, db, create_category, dummy_agent):
+    """Test if category exclusions work"""
+
+    # IMPORTANT: `dummy_agent` is actually needed here, since ``LiveSyncQueueEntry.create``
+    # will look for all existing agents in the DB. `db.session.add` is not necessary but
+    # it keeps static checkers happy.
+    db.session.add(dummy_agent)
```
Comment (ThiefMaster) [inline livesync/tests/queue_test.py]: You can use `@pytest.mark.usefixtures('dummy_agent')` if you just need to enable a fixture but not actually access it

File: livesync/indico_livesync/cli.py [hunk 10]
```diff
@@ -26,6 +26,7 @@
 from indico.modules.events.modes.events import Event
 from indico.util.console import cformat
 
+
```
Comment (ThiefMaster) [inline livesync/indico_livesync/cli.py]: too much whitespace

File: livesync/indico_livesync/uploader.py [hunk 11]
```diff
@@ -87,7 +87,6 @@ def processed_records(self, records):
             self.logger.debug('Marking as processed: %s', record)
             record.processed = True
         db.session.commit()
-        transaction.abort()  # clear ZEO cache
```
Comment (ThiefMaster) [inline livesync/indico_livesync/uploader.py]: Is the commit message correct? This doesn't seem to be related to livesync-invenio

File: livesync/indico_livesync/handler.py [hunk 12]
```diff
@@ -23,31 +23,31 @@
...
 
 def connect_signals(plugin):
     # request
     plugin.connect(signals.after_process, _apply_changes)
     plugin.connect(signals.before_retry, _clear_changes)
     # moved
-    # plugin.connect(signals.category.moved, _moved)
-    # plugin.connect(signals.event.moved, _moved)
+    plugin.connect(signals.category.moved, _moved)
+    plugin.connect(signals.event.moved, _moved_event)
```
Comment (ThiefMaster) [inline livesync/indico_livesync/handler.py]: `_event_moved` to be consistent with `_event_deleted`

File: livesync/indico_livesync/handler.py [hunk 13]
```diff
@@ -110,6 +107,10 @@ def _created(obj, **kwargs):
     _register_change(obj, ChangeType.created)
 
 
+def _event_deleted(obj, **kwargs):
```
Comment (ThiefMaster) [inline livesync/indico_livesync/handler.py]: I'd rename the first argument to `event` and decorate it with `@unify_event_args`. Less code to change once we change the signal to pass the new Event object (which we could do already - the only case where the legacy object is needed is the signal handler that logs the ZODB OID, and that could simply be moved into the Conference.delete method).

Actually, you could even use `_deleted` for the event deleted signal and add the decorator there.

File: livesync/indico_livesync/handler.py [hunk 14]
```diff
@@ -165,16 +137,14 @@ def _note_changed(note, **kwargs):
...
 
 
 def _apply_changes(sender, **kwargs):
     if not hasattr(g, 'livesync_changes'):
         return
     for ref, changes in g.livesync_changes.iteritems():
-        if is_ref_excluded(ref):
-            continue
-        LiveSyncQueueEntry.create(changes, ref)
+        LiveSyncQueueEntry.create(changes, ref, excluded_categories=get_excluded_categories())
```
Comment (ThiefMaster) [inline livesync/indico_livesync/handler.py]: does it really make sense to call it inside the loop? i know it's memoized, but it still seems cleaner to do it outside

File: livesync/indico_livesync/migrations/201607151736_205f944640f6_add_session_to_livesyncqueueentry.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+down_revision = '230c086da074'
+
+
+def upgrade():
+    op.add_column('queues', sa.Column('session_id', sa.Integer(), nullable=True), schema='plugin_livesync')
+    op.create_foreign_key(None,
+                          'queues', 'sessions',
+                          ['session_id'], ['id'],
+                          source_schema='plugin_livesync', referent_schema='events')
+    op.create_index(op.f('ix_queues_session_id'), 'queues', ['session_id'], unique=False, schema='plugin_livesync')
```
Comment (ThiefMaster) [inline livesync/indico_livesync/migrations/201607151736_205f944640f6_add_session_to_livesyncqueueentry.py]: first argument can be `None`; the name will be generated automatically

File: livesync/indico_livesync/util.py (hunk 8 above)
Comment (pferreir) [inline livesync/indico_livesync/util.py]: I think you originally wrote that some years ago 😉  But yeah, makes more sense.

File: livesync/tests/queue_test.py (hunk 9 above)
Comment (pferreir) [inline livesync/tests/queue_test.py]: Yeah, I didn't think of that. I had seen it before.

File: livesync/indico_livesync/handler.py (hunk 13 above)
Comment (pferreir) [inline livesync/indico_livesync/handler.py]: Jolly good idea! 👍

File: livesync/indico_livesync/handler.py (hunk 14 above)
Comment (pferreir) [inline livesync/indico_livesync/handler.py]: You mean `get_excluded_categories()`? No it doesn't.

Comment (ThiefMaster) [inline livesync/indico_livesync/handler.py]: yes, that one

File: livesync/indico_livesync/models/queue.py [hunk 16]
```diff
@@ -175,62 +209,61 @@ class LiveSyncQueueEntry(db.Model):
...
-            for change in changes:
+        obj = obj_deref(ref)
+
+        if isinstance(obj, Category):
+            if excluded_categories & {c.id for c in obj.chain_query}:
+                return
+        else:
+            event = obj if isinstance(obj, Event) else obj.event_new
+            if (Event.query
+                    .filter(Event.id == event.id, Event.category_chain_overlaps(excluded_categories))
```
Comment (ThiefMaster) [inline livesync/indico_livesync/models/queue.py]: wrong indentation

File: livesync/indico_livesync/models/queue.py [hunk 17]
```diff
@@ -175,62 +209,61 @@ class LiveSyncQueueEntry(db.Model):
...
+        :param excluded_categories: set of categories (IDs) whose items
+                                    will not be tracked
         """
         ref = dict(ref)
-        for agent in LiveSyncAgent.find():
-            for change in changes:
+        obj = obj_deref(ref)
+
+        if isinstance(obj, Category):
+            if excluded_categories & {c.id for c in obj.chain_query}:
```
Comment (ThiefMaster) [inline livesync/indico_livesync/models/queue.py]: `any(c.id in excluded_categories for c in obj.chain_query)`
